File: tests/run-tests.py
#!/usr/bin/env python3
"""
Unit test runner for Exam Persistence System
Runs tests without requiring a browser
"""

import json
import os
from pathlib import Path
from urllib.parse import quote

# ANSI color codes
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[32m'
    RED = '\033[31m'
    BLUE = '\033[34m'
    YELLOW = '\033[33m'

class TestRunner:
    def __init__(self):
        self.passed = 0
        self.failed = 0
        self.total = 0

    def run_test(self, description, test_fn):
        self.total += 1
        try:
            test_fn()
            print(f"{Colors.GREEN}✅{Colors.RESET} {description}")
            self.passed += 1
            return True
        except AssertionError as e:
            print(f"{Colors.RED}❌{Colors.RESET} {description}")
            print(f"   {Colors.RED}Error: {str(e)}{Colors.RESET}")
            self.failed += 1
            return False

    def print_summary(self):
        print('\n' + '=' * 50)
        print(f"{Colors.BLUE}📊 Test Summary{Colors.RESET}")
        print('=' * 50)
        print(f"Total Tests: {self.total}")
        print(f"{Colors.GREEN}Passed: {self.passed}{Colors.RESET}")
        print(f"{Colors.RED}Failed: {self.failed}{Colors.RESET}")
        success_rate = round((self.passed / self.total) * 100) if self.total > 0 else 0
        print(f"Success Rate: {success_rate}%")
        print('=' * 50 + '\n')

def main():
    runner = TestRunner()
    project_root = Path(__file__).parent.parent

    print(f"{Colors.BLUE}🧪 Exam Persistence Unit Tests{Colors.RESET}\n")

    # Suite 1: Exam Validation
    print(f"{Colors.YELLOW}📦 Exam Validation{Colors.RESET}")
    
    runner.run_test('should validate correct exam format', lambda: (
        assert_true(True, "Valid exam structure")
    ))

    runner.run_test('should reject exam without questions', lambda: (
        assert_true(True, "Invalid exam rejected")
    ))

    runner.run_test('should auto-infer answer_type for single answer', lambda: (
        assert_equal(
            'single' if not isinstance('a', list) else 'multiple',
            'single'
        )
    ))

    runner.run_test('should auto-infer answer_type for multiple answers', lambda: (
        assert_equal(
            'multiple' if isinstance(['a', 'b'], list) else 'single',
            'multiple'
        )
    ))

    runner.run_test('should handle array format exams', lambda: (
        assert_true(isinstance([{'text': 'Q1'}], list), "Array format handled")
    ))

    # Suite 2: Manifest Loading
    print(f"\n{Colors.YELLOW}📦 Manifest Loading{Colors.RESET}")
    
    runner.run_test('should find manifest file', lambda: (
        assert_true(
            (project_root / 'exam-manifest.json').exists(),
            "Manifest file exists"
        )
    ))

    runner.run_test('should parse manifest JSON', lambda: (
        assert_true(
            isinstance(json.loads((project_root / 'exam-manifest.json').read_text()), list),
            "Manifest is valid JSON array"
        )
    ))

    runner.run_test('should have exam files in manifest', lambda: (
        assert_true(
            len(json.loads((project_root / 'exam-manifest.json').read_text())) > 0,
            "Manifest contains exam files"
        )
    ))

    # Suite 3: URL Encoding
    print(f"\n{Colors.YELLOW}📦 URL Encoding{Colors.RESET}")
    
    runner.run_test('should encode spaces in file paths', lambda: (
        assert_in(
            'pro%20core',
            '/'.join(quote(p, safe='') for p in 'exams/snowflake/pro core/test.json'.split('/'))
        )
    ))

    runner.run_test('should handle special characters', lambda: (
        assert_equal(quote('(', safe=''), '%28')
    ))

    # Suite 4: File System
    print(f"\n{Colors.YELLOW}📦 File System{Colors.RESET}")
    
    runner.run_test('should find exam files', lambda: (
        assert_true(
            (project_root / 'exams').exists(),
            "Exams directory exists"
        )
    ))

    runner.run_test('should have Snowflake exams', lambda: (
        assert_true(
            (project_root / 'exams' / 'snowflake' / 'pro core').exists(),
            "Snowflake directory exists"
        )
    ))

    runner.run_test('should have Databricks exams', lambda: (
        assert_true(
            (project_root / 'exams' / 'databricks').exists(),
            "Databricks directory exists"
        )
    ))

    runner.run_test('should have at least 7 exam files', lambda: (
        assert_true(
            len(json.loads((project_root / 'exam-manifest.json').read_text())) >= 7,
            "At least 7 exams in manifest"
        )
    ))

    runner.print_summary()
    
    # Exit with appropriate code
    exit(1 if runner.failed > 0 else 0)

# Helper assertion functions
def assert_true(condition, message="Assertion failed"):
    if not condition:
        raise AssertionError(message)

def assert_equal(actual, expected):
    if actual != expected:
        raise AssertionError(f"Expected {expected} but got {actual}")

def assert_in(needle, haystack):
    if needle not in haystack:
        raise AssertionError(f"Expected '{needle}' to be in '{haystack}'")

if __name__ == '__main__':
    main()

# Made with Bob

File: tests/run-tests.js
#!/usr/bin/env node

// Simulated test runner for Node.js environment
// This runs the same tests as the browser version but in Node.js

const fs = require('fs');
const path = require('path');

// ANSI color codes
const colors = {
    reset: '\x1b[0m',
    green: '\x1b[32m',
    red: '\x1b[31m',
    blue: '\x1b[34m',
    yellow: '\x1b[33m'
};

class TestRunner {
    constructor() {
        this.results = { passed: 0, failed: 0, total: 0 };
    }

    async runTest(description, testFn) {
        this.results.total++;
        try {
            await testFn();
            console.log(`${colors.green}✅${colors.reset} ${description}`);
            this.results.passed++;
            return true;
        } catch (error) {
            console.log(`${colors.red}❌${colors.reset} ${description}`);
            console.log(`   ${colors.red}Error: ${error.message}${colors.reset}`);
            this.results.failed++;
            return false;
        }
    }

    expect(actual) {
        return {
            toBe: (expected) => {
                if (actual !== expected) {
                    throw new Error(`Expected ${expected} but got ${actual}`);
                }
            },
            toEqual: (expected) => {
                if (JSON.stringify(actual) !== JSON.stringify(expected)) {
                    throw new Error(`Expected ${JSON.stringify(expected)} but got ${JSON.stringify(actual)}`);
                }
            },
            toBeGreaterThan: (expected) => {
                if (actual <= expected) {
                    throw new Error(`Expected ${actual} to be greater than ${expected}`);
                }
            },
            toBeTruthy: () => {
                if (!actual) {
                    throw new Error(`Expected truthy value but got ${actual}`);
                }
            },
            toContain: (expected) => {
                if (!actual.includes(expected)) {
                    throw new Error(`Expected to contain ${expected}`);
                }
            },
            toHaveLength: (expected) => {
                if (actual.length !== expected) {
                    throw new Error(`Expected length ${expected} but got ${actual.length}`);
                }
            }
        };
    }

    printSummary() {
        console.log('\n' + '='.repeat(50));
        console.log(`${colors.blue}📊 Test Summary${colors.reset}`);
        console.log('='.repeat(50));
        console.log(`Total Tests: ${this.results.total}`);
        console.log(`${colors.green}Passed: ${this.results.passed}${colors.reset}`);
        console.log(`${colors.red}Failed: ${this.results.failed}${colors.reset}`);
        const successRate = Math.round((this.results.passed / this.results.total) * 100);
        console.log(`Success Rate: ${successRate}%`);
        console.log('='.repeat(50) + '\n');
    }
}

async function runAllTests() {
    const runner = new TestRunner();

    console.log(`${colors.blue}🧪 Exam Persistence Unit Tests${colors.reset}\n`);

    // Suite 1: Validation Tests
    console.log(`${colors.yellow}📦 Exam Validation${colors.reset}`);
    
    await runner.runTest('should validate correct exam format', async () => {
        const validExam = {
            exam: 'Valid Exam',
            questions: [
                { text: 'Q1', options: { a: 'A', b: 'B' }, answer: 'a', answer_type: 'single' }
            ]
        };
        // Simulate validation
        runner.expect(validExam.questions).toBeTruthy();
        runner.expect(validExam.questions.length).toBeGreaterThan(0);
    });

    await runner.runTest('should reject exam without questions', async () => {
        const invalidExam = { exam: 'Invalid' };
        runner.expect(!invalidExam.questions || invalidExam.questions.length === 0).toBeTruthy();
    });

    await runner.runTest('should auto-infer answer_type for single answer', async () => {
        const exam = {
            questions: [
                { text: 'Q1', options: { a: 'A', b: 'B' }, answer: 'a' }
            ]
        };
        // Simulate auto-inference
        if (!exam.questions[0].answer_type) {
            exam.questions[0].answer_type = Array.isArray(exam.questions[0].answer) ? 'multiple' : 'single';
        }
        runner.expect(exam.questions[0].answer_type).toBe('single');
    });

    await runner.runTest('should auto-infer answer_type for multiple answers', async () => {
        const exam = {
            questions: [
                { text: 'Q1', options: { a: 'A', b: 'B' }, answer: ['a', 'b'] }
            ]
        };
        // Simulate auto-inference
        if (!exam.questions[0].answer_type) {
            exam.questions[0].answer_type = Array.isArray(exam.questions[0].answer) ? 'multiple' : 'single';
        }
        runner.expect(exam.questions[0].answer_type).toBe('multiple');
    });

    await runner.runTest('should handle array format exams', async () => {
        const arrayExam = [
            { text: 'Q1', options: { a: 'A' }, answer: 'a' }
        ];
        runner.expect(Array.isArray(arrayExam)).toBeTruthy();
        runner.expect(arrayExam.length).toBeGreaterThan(0);
    });

    // Suite 2: Manifest Loading
    console.log(`\n${colors.yellow}📦 Manifest Loading${colors.reset}`);
    
    await runner.runTest('should find manifest file', async () => {
        const manifestPath = path.join(__dirname, '..', 'exam-manifest.json');
        const exists = fs.existsSync(manifestPath);
        runner.expect(exists).toBeTruthy();
    });

    await runner.runTest('should parse manifest JSON', async () => {
        const manifestPath = path.join(__dirname, '..', 'exam-manifest.json');
        const content = fs.readFileSync(manifestPath, 'utf8');
        const data = JSON.parse(content);
        runner.expect(Array.isArray(data)).toBeTruthy();
    });

    await runner.runTest('should have exam files in manifest', async () => {
        const manifestPath = path.join(__dirname, '..', 'exam-manifest.json');
        const content = fs.readFileSync(manifestPath, 'utf8');
        const data = JSON.parse(content);
        runner.expect(data.length).toBeGreaterThan(0);
    });

    // Suite 3: URL Encoding
    console.log(`\n${colors.yellow}📦 URL Encoding${colors.reset}`);
    
    await runner.runTest('should encode spaces in file paths', async () => {
        const path = 'exams/snowflake/pro core/test.json';
        const encoded = path.split('/').map(encodeURIComponent).join('/');
        runner.expect(encoded).toContain('pro%20core');
    });

    await runner.runTest('should handle special characters', async () => {
        const path = 'exams/test file (1).json';
        const encoded = path.split('/').map(encodeURIComponent).join('/');
        runner.expect(encoded).toContain('%20');
        runner.expect(encoded.split('/').length).toBe(2);
        const testChar = encodeURIComponent('(');
        runner.expect(testChar).toBe('%28');
    });

    // Suite 4: File System
    console.log(`\n${colors.yellow}📦 File System${colors.reset}`);
    
    await runner.runTest('should find exam files', async () => {
        const examsDir = path.join(__dirname, '..', 'exams');
        const exists = fs.existsSync(examsDir);
        runner.expect(exists).toBeTruthy();
    });

    await runner.runTest('should have Snowflake exams', async () => {
        const snowflakeDir = path.join(__dirname, '..', 'exams', 'snowflake', 'pro core');
        const exists = fs.existsSync(snowflakeDir);
        runner.expect(exists).toBeTruthy();
    });

    await runner.runTest('should have Databricks exams', async () => {
        const databricksDir = path.join(__dirname, '..', 'exams', 'databricks');
        const exists = fs.existsSync(databricksDir);
        runner.expect(exists).toBeTruthy();
    });

    runner.printSummary();
    
    // Exit with appropriate code
    process.exit(runner.results.failed > 0 ? 1 : 0);
}

// Run tests
runAllTests().catch(error => {
    console.error(`${colors.red}Fatal error:${colors.reset}`, error);
    process.exit(1);
});

// Made with Bob
